### PR TITLE
Services: Add MCU

### PR DESCRIPTION
**Pull request by valentinvanelslande**
Implement Set3DLEDState
Stub GetBatteryLevel and GetSoundVolume (HWC & HID)

------------------------------

This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3452)

### CODE REVIEW

File: src/core/hle/service/mcu/cdc.cpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+// Refer to the license.txt file included.
+
+#include "core/hle/service/mcu/cdc.h"
+
+namespace Service {
+namespace MCU {
+
+CDC::CDC(std::shared_ptr<Module> mcu) : Module::Interface(std::move(mcu), "mcu::CDC", 1) {
+    static const FunctionInfo functions[] = {
+        {0x00010000, nullptr, "unk_0x00010000"},
```
Comment (mailwl) [inline src/core/hle/service/mcu/cdc.cpp]: if i am right, unknown functions not listed in `functions`

**Comment by chris062689** [comment]
What is this change actually doing? What reverse engineering has been done? This PR needs more information instead of an empty description.

**Comment by wwylele** [comment]
Does any game/application directly calls MCU functions? If not, it only make sense to stub/implement them when we are going to LLE some system module. Currently this looks just like a bunch of low effort code that nobody / no application is going to touch for a long time.

**Comment by valentinvanelslande** [comment]
@chris062689 edited description.
No reverse engineering was done
Fixes [3DSident] (https://github.com/joel16/3DSident) and MCUinfo bugs
Now 3DSident shows valid values in Battery information and MCUinfo shows 0.16 MCU firmware

3DSident Battery information (master):
3dsident-master

3DSident Battery information (PR):
3dsident-pr


MCUinfo (master):
mcuinfo-master

MCUinfo (PR):
mcuinfo-pr

File: src/core/hle/service/mcu/mcu.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+#include "core/hle/service/mcu/pls.h"
+#include "core/hle/service/mcu/rtc.h"
+#include "core/hle/service/mcu/snd.h"
+#include "core/hle/shared_page.h"
+
+namespace Service {
+namespace MCU {
+
+void Module::Interface::GetBatteryLevel(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x000500, 0, 0);
```
Comment (wwylele) [inline src/core/hle/service/mcu/mcu.cpp]: The main ID is `0x0005`, not `0x000500`

File: src/core/hle/service/mcu/mcu.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
```
Comment (wwylele) [inline src/core/hle/service/mcu/mcu.cpp]: You pushed two words below, so this should likely be `(2, 0)`

File: src/core/hle/service/mcu/mcu.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    rb.Push(RESULT_SUCCESS);
+    rb.Push<u8>(0x64);
+    LOG_WARNING(Service_MCU, "(STUBBED) called");
+}
+
+void Module::Interface::Set3DLEDState(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x000900, 4, 0);
```
Comment (wwylele) [inline src/core/hle/service/mcu/mcu.cpp]: `(0x0009, 1, 0)`. I will stop commenting on header codes (They are basically all wrong)

File: src/core/hle/service/mcu/hid.cpp [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+namespace Service {
+namespace MCU {
+
+HID::HID(std::shared_ptr<Module> mcu) : Module::Interface(std::move(mcu), "mcu::HID", 1) {
+    static const FunctionInfo functions[] = {
+        {0x00060000, nullptr, "ReadGyroscopeValues"},
+        {0x00070000, nullptr, "GetRaw3DSliderPosition"},
+        {0x000C0000, nullptr, "GetMcuHidEventHandle"},
+        {0x000D0000, nullptr, "GetMcuHidEventReason"},
+        {0x000E0000, &HID::GetSoundVolume2, "GetSoundVolume"},
```
Comment (mailwl) [inline src/core/hle/service/mcu/hid.cpp]: any reason not to move GetSoundVolume2 to HID class?

**Comment by valentinvanelslande** [comment]
@wwylele I have the debug build, now how I check all the DEBUG_ASSERT in the IPC translation?

**Comment by wwylele** [comment]
@valentinvanelslande Just run something that calls the function...

**Comment by valentinvanelslande** [comment]
works

File: src/core/hle/service/mcu/mcu.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    IPC::RequestParser rp(ctx, 0x0009, 1, 0);
+    u8 state = rp.Pop<u8>();
+    SharedPage::shared_page.ledstate_3d = state;
+    IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
+    rb.Push(RESULT_SUCCESS);
+}
+
+void Module::Interface::GetSoundVolume(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x000B, 0, 0);
+    IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
```
Comment (wwylele) [inline src/core/hle/service/mcu/mcu.cpp]: You still didn't fix this. You clearly push two words below so this should be 2, 0. Same for GetSoundVolume2

File: src/core/hle/service/mcu/mcu.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+         *  Outputs:
+         *      1 : Result of function, 0 on success, otherwise error code
+         */
+        void Set3DLEDState(Kernel::HLERequestContext& ctx);
+
+        /**
+         * MCU::GetSoundVolume service function
+         *  Outputs:
+         *      1 : Result of function, 0 on success, otherwise error code
+         *      2 : Sound volume
```
Comment (jroweboy) [inline src/core/hle/service/mcu/mcu.h]: No mention of volume range (which iirc on 3dbrew its 0 - 0x3F ?)

File: src/core/hle/service/mcu/mcu.h [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+public:
+    class Interface : public ServiceFramework<Interface> {
+    public:
+        Interface(std::shared_ptr<Module> mcu, const char* name, u32 max_session);
+
+        /**
+         * MCU::GetBatteryLevel service function
+         *  Outputs:
+         *      1 : Result of function, 0 on success, otherwise error code
+         *      2 : Battery level (0-64)
```
Comment (jroweboy) [inline src/core/hle/service/mcu/mcu.h]: This comment says from 0 - 64 but when you actually return the value its 0x64 and 0x64 != 64 :P

File: src/core/hle/service/mcu/mcu.cpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    rb.Push(RESULT_SUCCESS);
+    rb.Push<u8>(0x64);
+
+    NGLOG_WARNING(Service_MCU, "(STUBBED) called");
+}
+
+void Module::Interface::Set3DLEDState(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x0009, 1, 0);
+    u8 state = rp.Pop<u8>();
+    SharedPage::shared_page.ledstate_3d = state;
```
Comment (wwylele) [inline src/core/hle/service/mcu/mcu.cpp]: I can't find the documentation stating that this service function controls this shared page field. Is this RE-ed by yourself, or could you provide the reference?

File: src/core/hle/service/mcu/mcu.cpp [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+#include "core/hle/shared_page.h"
+
+namespace Service {
+namespace MCU {
+
+void Module::Interface::GetBatteryLevel(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x0005, 0, 0);
+    IPC::RequestBuilder rb = rp.MakeBuilder(2, 0);
+    rb.Push(RESULT_SUCCESS);
+    rb.Push<u8>(0x64);
```
Comment (wwylele) [inline src/core/hle/service/mcu/mcu.cpp]: IMO putting `100` here is more intuitive.

**Comment by wwylele** [comment]
Closing due to inactivity

**Pull request closed without merging**